pkg/controller/pxc/upgrade.go

```diff
@@ -191,41 +192,80 @@ func (r *ReconcilePerconaXtraDBCluster) smartUpdate(sfs api.StatefulApp, cr *api
 			primaryPod = pod
 		} else {
 			log.Info(fmt.Sprintf("apply changes to secondary pod %s", pod.Name))
-			if err := r.applyNWait(cr, sfs.StatefulSet().Status.UpdateRevision, &pod, waitLimit); err != nil {
+			if err := r.applyNWait(cr, sfs.StatefulSet(), &pod, waitLimit); err != nil {
 				return fmt.Errorf("failed to apply changes: %v", err)
 			}
 		}
 	}
 
 	log.Info(fmt.Sprintf("apply changes to primary pod %s", primaryPod.Name))
-	if err := r.applyNWait(cr, sfs.StatefulSet().Status.UpdateRevision, &primaryPod, waitLimit); err != nil {
+	if err := r.applyNWait(cr, sfs.StatefulSet(), &primaryPod, waitLimit); err != nil {
 		return fmt.Errorf("failed to apply changes: %v", err)
 	}
 
 	return nil
 }
 
-func (r *ReconcilePerconaXtraDBCluster) applyNWait(cr *api.PerconaXtraDBCluster, updateRevision string, pod *corev1.Pod, waitLimit int) error {
-	if pod.ObjectMeta.Labels["controller-revision-hash"] == updateRevision {
+func (r *ReconcilePerconaXtraDBCluster) applyNWait(cr *api.PerconaXtraDBCluster, sfs *appsv1.StatefulSet, pod *corev1.Pod, waitLimit int) error {
+	if pod.ObjectMeta.Labels["controller-revision-hash"] == sfs.Status.UpdateRevision {
 		log.Info(fmt.Sprintf("pod %s is already updated", pod.Name))
 	} else {
 		if err := r.client.Delete(context.TODO(), pod); err != nil {
 			return fmt.Errorf("failed to delete pod: %v", err)
 		}
 	}
 
-	if err := r.waitPodRestart(updateRevision, pod, waitLimit); err != nil {
+	if err := r.waitPodRestart(sfs.Status.UpdateRevision, pod, waitLimit); err != nil {
 		return fmt.Errorf("failed to wait pod: %v", err)
 	}
 
+	if err := r.waitUntilOnline(cr, pod, waitLimit); err != nil {
+		return fmt.Errorf("failed to wait pxc status: %v", err)
+	}
+
 	if err := r.waitPXCSynced(cr, pod.Status.PodIP, waitLimit); err != nil {
 		return fmt.Errorf("failed to wait pxc sync: %v", err)
 	}
 
 	return nil
 }
 
-func (r *ReconcilePerconaXtraDBCluster) getPrimaryPod(cr *api.PerconaXtraDBCluster) (string, error) {
+func (r *ReconcilePerconaXtraDBCluster) waitUntilOnline(cr *api.PerconaXtraDBCluster, pod *corev1.Pod, waitLimit int) error {
+	database, err := r.proxyDB(cr)
+	if err != nil {
+		return fmt.Errorf("failed to get proxySQL db: %v", err)
+	}
+
+	defer database.Close()
+
+	for i := 0; i < waitLimit; i++ {
+		statuses, err := database.Status(pod.Name, pod.Status.PodIP)
```

Comment:
could we pass also sts?
reason `like "pod-1%"` can select `pod-11`
```suggestion
		statuses, err := database.Status(pod.Name + "." + sfsName, pod.Status.PodIP)
```

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/457/commits/3478c72d4720b97eacb06ed77c5c4c9a3f2cfd14